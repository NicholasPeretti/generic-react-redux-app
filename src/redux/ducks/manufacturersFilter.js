export const NAMESPACE = 'MANUFACTURERS_FILTER'
export const SET_FETCHING = `${NAMESPACE} SET_FETCHING`
export const SET_ERROR = `${NAMESPACE} SET_ERROR`
export const SET_MANUFACTURERS = `${NAMESPACE} SET_MANUFACTURERS`
export const SELECT_MANUFACTURER = `${NAMESPACE} SELECT_MANUFACTURER`

export const selectManufacturer = manufacturer => ({
  type: SELECT_MANUFACTURER,
  manufacturer
})

export const setManufacturers = manufacturers => ({
  type: SET_MANUFACTURERS,
  manufacturers
})

export const setError = error => ({
  type: SET_ERROR,
  error
})

export const setFetching = fetching => ({
  type: SET_FETCHING,
  fetching
})

export const defaultState = {
  fetching: false,
  error: null,
  manufacturers: [],
  selectedManufacturer: null
}

export default function reducer (state = defaultState, action = {}) {
  switch (action.type) {
    case SET_MANUFACTURERS: {
      state = Object.assign({}, state, {
        manufacturers: action.manufacturers.map(el => el.name)
      })
      break
    }

    case SELECT_MANUFACTURER: {
      state = Object.assign({}, state, {
        selectedManufacturer: action.manufacturer
      })
      break
    }

    case SET_ERROR: {
      state = Object.assign({}, state, {
        error: action.error
      })
      break
    }

    case SET_FETCHING: {
      state = Object.assign({}, state, {
        fetching: action.fetching
      })
      break
    }
  }

  return state
}
