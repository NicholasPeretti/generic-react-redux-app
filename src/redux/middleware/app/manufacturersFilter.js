import { setFetching, setError, setManufacturers, NAMESPACE } from '../../ducks/manufacturersFilter'
import { isApiAction, API_ERROR, API_REQUEST, API_SUCCESS, apiRequest } from '../core/api'
import { GET_MANUFACTURERS } from '../../endpoints'

export const fetchManufacturers = () => {
  const { url, method } = GET_MANUFACTURERS
  return apiRequest(url, method, NAMESPACE)
}

export default function middleware () {
  return dispatch => action => {
    dispatch(action)

    handleApiMiddlewareActions(dispatch, action)
  }
}

function handleApiMiddlewareActions (dispatch, action) {
  if (!isApiAction(action, NAMESPACE)) return

  switch (action.type) {
    case API_REQUEST: {
      dispatch(setFetching(true))
      dispatch(setError(null))
      break
    }

    case API_SUCCESS: {
      const { manufacturers } = action.payload
      dispatch(setManufacturers(manufacturers))
      dispatch(setFetching(false))
      break
    }

    case API_ERROR: {
      const error = action.payload
      dispatch(setError(error))
      dispatch(setFetching(false))
      break
    }
  }
}
