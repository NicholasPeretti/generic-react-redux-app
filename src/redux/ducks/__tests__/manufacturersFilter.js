import DuckTest from '../../../utils/tests/duck'
import * as manufacturersFilter from '../manufacturersFilter'
import { Manufacturers as MOCK_MANUFACTURERS } from '../../../utils/tests/data'

describe('ManufacturersFilter', () => {
  DuckTest(manufacturersFilter)

  test('defaultState should match snapshot', () => {
    expect(manufacturersFilter.defaultState).toMatchSnapshot()
  })

  describe('Action creators', () => {
    describe('setManufacturers', () => {
      test('Should have the correct type', () => {
        const action = manufacturersFilter.setManufacturers()
        expect(action.type).toEqual(manufacturersFilter.SET_MANUFACTURERS)
      })

      test('Should have the correct value', () => {
        const action = manufacturersFilter.setManufacturers(MOCK_MANUFACTURERS)
        expect(action.manufacturers).toEqual(MOCK_MANUFACTURERS)
      })
    })

    describe('selectManufacturer', () => {
      test('Should have the correct type', () => {
        const manufacturer = MOCK_MANUFACTURERS[0].name
        const action = manufacturersFilter.selectManufacturer(manufacturer)
        expect(action.type).toEqual(manufacturersFilter.SELECT_MANUFACTURER)
      })

      test('Should have the correct value', () => {
        const manufacturer = MOCK_MANUFACTURERS[0].name
        const action = manufacturersFilter.selectManufacturer(manufacturer)
        expect(action.manufacturer).toEqual(manufacturer)
      })
    })

    describe('setError', () => {
      test('Should have the correct type', () => {
        const action = manufacturersFilter.setError()
        expect(action.type).toEqual(manufacturersFilter.SET_ERROR)
      })

      test('Should have the correct value', () => {
        const error = 'Error'
        const action = manufacturersFilter.setError(error)
        expect(action.error).toEqual(error)
      })
    })

    describe('setFetching', () => {
      test('Should have the correct type', () => {
        const action = manufacturersFilter.setFetching(true)
        expect(action.type).toEqual(manufacturersFilter.SET_FETCHING)
      })

      test('Should have the correct value', () => {
        const action = manufacturersFilter.setFetching(true)
        expect(action.fetching).toEqual(true)
      })
    })
  })

  describe('Reducer', () => {
    describe('setError', () => {
      test('Should set the `error` property', () => {
        const error = 'Error'
        const newState = manufacturersFilter.default(
          manufacturersFilter.defaultState,
          manufacturersFilter.setError(error)
        )

        expect(newState.error).toEqual(error)
      })
    })

    describe('setFetching', () => {
      test('Should set the `fetching` property', () => {
        const newState = manufacturersFilter.default(
          manufacturersFilter.defaultState,
          manufacturersFilter.setFetching(true)
        )

        expect(newState.fetching).toEqual(true)
      })
    })

    describe('setManufacturers', () => {
      test('Should set the `manufacturers` property', () => {
        const newState = manufacturersFilter.default(
          manufacturersFilter.defaultState,
          manufacturersFilter.setManufacturers(MOCK_MANUFACTURERS)
        )

        const expectedValue = MOCK_MANUFACTURERS.map(val => val.name)

        expect(newState.manufacturers).toEqual(expectedValue)
      })
    })

    describe('selectManufacturer', () => {
      test('Should set the `selectedManufacturer` property', () => {
        const manufacturer = MOCK_MANUFACTURERS[0].name
        const newState = manufacturersFilter.default(
          manufacturersFilter.defaultState,
          manufacturersFilter.selectManufacturer(manufacturer)
        )

        expect(newState.selectedManufacturer).toEqual(manufacturer)
      })
    })
  })
})
