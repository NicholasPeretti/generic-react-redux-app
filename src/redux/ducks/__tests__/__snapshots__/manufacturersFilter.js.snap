// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManufacturersFilter defaultState should match snapshot 1`] = `
Object {
  "error": null,
  "fetching": false,
  "manufacturers": Array [],
  "selectedManufacturer": null,
}
`;
